### PR TITLE
[FLINK-1775] BarrierBuffer fix to avoid end of stream deadlock

### PR DESCRIPTION
**Pull request by gyfora**
This fixes several corner-cases in the barrier buffer logic.

### CODE REVIEW

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/streamvertex/StreamVertex.java [hunk 1]
```diff
@@ -283,16 +288,18 @@ public StreamConfig getConfig() {
...
+		if (this.isRunning) {
+			try {
+				outputHandler.broadcastBarrier(id);
+				// TODO checkpoint state here
+				confirmBarrier(id);
+				if (LOG.isDebugEnabled()) {
+					LOG.debug("Superstep " + id + " processed: " + StreamVertex.this);
+				}
+			} catch (Exception e) {
+				// TODO:Figure this out properly
```
Comment (rmetzger) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/streamvertex/StreamVertex.java]: Why are you swallowing the exception here silently?

**Comment by uce** [comment]
Hey gyula,

What is the problem with the stream writer? And how can I reproduce the record writer broadcast problem? If you give me some instructions to reproduce the problem, I could also look into it.

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/streamvertex/StreamVertex.java (hunk 1 above)
Comment (gyfora) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/streamvertex/StreamVertex.java]: I had some weird exceptions after the streams finished, and in any case this shouldnt affect the recovery logic

Comment (gyfora) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/streamvertex/StreamVertex.java]: but I am trying to sort all these out with ufuk

**Comment by gyfora** [comment]
Should we merge this anyway? The recordwriter issue seems to be independent from this, and this commit solves some other bugs.

**Comment by gyfora** [comment]
Now it seems that all these issues are fixed.

**Comment by uce** [comment]
I think we need to refactor some parts, but this is a crucial fix. Tests are passing.

+1 to merge after throwing the Exception Robert mentioned.

**Comment by gyfora** [comment]
Ok I will add that one

**Pull request merged**